ee/server/service/users.go

```diff
@@ -0,0 +1,38 @@
+package service
+
+import (
+	"context"
+	"errors"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/fleetdm/fleet/v4/server/ptr"
+)
+
+// GetSSOUser is the premium implementation of svc.GetSSOUser, it allows to
+// create users during the SSO flow the first time they log in if
+// config.SSOSettings.EnableJITProvisioning is `true`
+func (svc *Service) GetSSOUser(ctx context.Context, auth fleet.Auth) (*fleet.User, error) {
+	config, err := svc.ds.AppConfig(ctx)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "getting app config")
+	}
+
+	user, err := svc.Service.GetSSOUser(ctx, auth)
+	var nfe fleet.NotFoundError
+	if !errors.As(err, &nfe) || !config.SSOSettings.EnableJITProvisioning {
+		return user, err
+	}
+
+	user, err = svc.Service.NewUser(ctx, fleet.UserPayload{
+		Name:       ptr.String(auth.UserDisplayName()),
```

Comment:
Is there a chance for `auth.UserID()`, `auth.UserDisplayName()` to be empty? Btw, we should document the `fleet.Auth` type:
https://github.com/fleetdm/fleet/blob/f47497d097dc7271217823bf129cc2cbe009e54f/server/fleet/sessions.go#L7-L11

Reply:
> Is there a chance for auth.UserID(), auth.UserDisplayName() to be empty?

- `auth.UserID`, technically no, but now that I think of it if the provider integration is misconfigured it might not be an email. I will look if we have email validation set up somewhere
- `auth.UserDisplayName` can be empty, but I looked through the codebase and seems like that would be OK, from the [Epic](https://github.com/fleetdm/fleet/issues/2008):

  > Fleet pulls the email and full name from SSO and applies it to the new user. If there is no full name, just apply email.

Reply:
so, seems like we don't do any kind of server-side validation for emails:

```
curl 'https://localhost:8080/api/latest/fleet/users/admin' -X POST -H 'Authorization: Bearer $TOKEN' --data-raw '{"email":"asd","name":"asdf","password":"as;lkdfjasdlk;fja3234@","global_role":"observer","teams":[]}'
{
  "user": {
    "created_at": "0001-01-01T00:00:00Z",
    "updated_at": "0001-01-01T00:00:00Z",
    "id": 25,
    "name": "asdf",
    "email": "asd",
    "force_password_reset": true,
    "gravatar_url": "",
    "sso_enabled": false,
    "global_role": "observer",
    "api_only": false,
    "teams": []
  }
}
```

should I create an issue and follow up, or do you think adding some kind of email validation should be part of this work?

Reply:
note: I added a sanity check for an empty `auth.UserID()`, just in case

Reply:
note II: I added `auth.UserID` as a fallback for the name.

the only thing left to define is if we want email validation or not.